### ISSUE TITLE
systemd-analyze  plot generates weird plots when systemd-based initrd in  use

### ISSUE THREAD

**Issue by arianvp** [systemd/systemd#29010]
### systemd version the issue has been seen with

253.6

### Used distribution

NixOS 23.05

### Linux kernel version used

6.1.46

### CPU architectures issue was seen on

aarch64

### Component

systemd-analyze

### Expected behaviour you didn't see

I expected all the services that run in initrd to show up in  the plot or non at all.

### Unexpected behaviour you saw

Instead  I just see a small subset of  `.service`s  that run in initrd and a lot of them are missing.

Furthermore `.device` units show up but systemd thinks they took the  entirety of the time spent in initrd  to  activate whilst I know this is not true. (e.g. `dev-dm\x2d0.device` shows up as activated before `dev-vda3` and `dev-vda2` finish activating; which is physically impossible).

Also these `.device` units now clobber `systemd-analyze blame` as well
```
$ systemd-analyze blame
1.677s dev-vport1p0.device
1.677s sys-devices-pci0000:00-0000:00:05.0-virtio1-virtio\x2dports-vport1p0.dev>
1.664s dev-hvc1.device
1.664s sys-devices-virtual-tty-hvc1.device
1.663s dev-hvc2.device
1.663s sys-devices-virtual-tty-hvc2.device
1.663s sys-devices-virtual-tty-hvc5.device
1.663s dev-hvc5.device
1.663s dev-hvc4.device
1.663s sys-devices-virtual-tty-hvc4.device
1.662s sys-devices-virtual-tty-hvc3.device
1.662s dev-hvc3.device
1.662s dev-ttyS0.device
1.662s sys-devices-platform-serial8250-tty-ttyS0.device
1.662s dev-ttyS3.device
1.662s sys-devices-platform-serial8250-tty-ttyS3.device
(...sic...)
```

![plot](https://gist.githubusercontent.com/arianvp/5e93279441c0265481e47779831e0764/raw/09472022c8680b1854befb724a612b2b11943ce8/plot.svg)

### Steps to reproduce the problem

```
systemd-analyze plot
```

**Comment by arianvp** [comment]
Arch Linux seems to have more services showing up in initrd. But is also missing some.
e.g. `systemd-udevd` seems to be missing in initrd.

Also arch has the same issues with the `.device` units having longer activation times than they actually did (they were already activated in stage 1)


![img](https://gist.githubusercontent.com/arianvp/8f9f46ba2783ed803282dd69ec32b7c4/raw/e4669d0f4937584749f86a1aa97fced9d1192af8/plot.svg)

**Comment by Winterhuman** [comment]
The `.device` duplication was reported at https://github.com/systemd/systemd/issues/23629, it's worth noting the device units are [processed in parallel](https://github.com/systemd/systemd/issues/23629#issuecomment-1184368806) so that *might* be why certain units appear to be finishing before their dependent devices

**Comment by arianvp** [comment]
I'm curious if this might be due to units being garbage collected before the transition to stage-2 or something.

Maybe systemd-analyze doesn't really support this use case

**Comment by yuwata** [comment]
> e.g. `systemd-udevd` seems to be missing in initrd part of the plot

If a service run both in initrd and the host system, then the timestamps in initrd are lost when the service started in the host.

**Comment by lf-** [comment]
Yeah, if you look at a plot next to the log, e.g.: you find that the units are just marked started when systemd in the main system actually starts.

This is mostly uninteresting, but I find it very frustrating that systemd-analyze plot doesn't give me useful info about my initrd as I then need to manually stare at it (which, it is 3 seconds, that's pretty sad).

```
[    3.260578] nixos systemd[1]: Activating default unit: default.target
[    3.260642] nixos systemd[1]: multi-user.target: Trying to enqueue job multi-user.target/start/isolate
[    3.260702] nixos systemd[1]: time-set.target: Cannot add dependency job, ignoring: Unit time-set.target not found.
[    3.260752] nixos systemd[1]: tmp.mount: Cannot add dependency job, ignoring: Unit tmp.mount not found.
[    3.260776] nixos systemd[1]: first-boot-complete.target: Cannot add dependency job, ignoring: Unit first-boot-complete.target not found.
```

https://gist.github.com/lf-/4bd10813c2851398a12d3d607685bbd7
